hpedockerplugin/hpe/hpe_3par_common.py

```diff
@@ -938,3 +967,199 @@ def _get_prioritized_host_on_3par(self, host, hosts, hostname):
                 return host, hostname
 
         return host, hostname
+
+    def create_snapshot(self, snapshot):
+        LOG.debug("Create Snapshot\n%s", json.dumps(snapshot, indent=2))
+
+        try:
+            snap_name = self._get_3par_snap_name(snapshot['id'])
+            vol_name = self._get_3par_vol_name(snapshot['volume_id'])
+
+            extra = {'volume_name': snapshot['volume_name']}
+            vol_id = snapshot.get('volume_id', None)
+            if vol_id:
+                extra['volume_id'] = vol_id
+
+            try:
+                extra['display_name'] = snapshot['display_name']
+            except AttributeError:
+                pass
+
+            try:
+                extra['description'] = snapshot['display_description']
+            except AttributeError:
+                pass
+
+            optional = {'comment': json.dumps(extra),
+                        'readOnly': True}
+            if snapshot['expirationHours']:
+                optional['expirationHours'] = snapshot['expirationHours']
+            if snapshot['retentionHours']:
+                optional['retentionHours'] = snapshot['retentionHours']
+
+            self.client.createSnapshot(snap_name, vol_name, optional)
+        except hpeexceptions.HTTPForbidden as ex:
+            LOG.error("Exception: %s", ex)
+            raise exception.NotAuthorized()
+        except hpeexceptions.HTTPNotFound as ex:
+            LOG.error("Exception: %s", ex)
+            raise exception.NotFound()
+
+    def create_cloned_volume(self, dst_volume, src_vref):
+        try:
+            dst_3par_vol_name = self._get_3par_vol_name(dst_volume['id'])
+            src_3par_vol_name = self._get_3par_vol_name(src_vref['id'])
+            # back_up_process = False
+            vol_chap_enabled = False
+
+            # Check whether a volume is ISCSI and CHAP enabled on it.
+            if self.config.hpe3par_iscsi_chap_enabled:
+                try:
+                    vol_chap_enabled = self.client.getVolumeMetaData(
+                        src_3par_vol_name, 'HPQ-docker-CHAP-name')['value']
+                except hpeexceptions.HTTPNotFound:
+                    LOG.debug("CHAP is not enabled on volume %(vol)s ",
+                              {'vol': src_vref['id']})
+                    vol_chap_enabled = False
+
+            # if the sizes of the 2 volumes are the same and except backup
+            # process for ISCSI volume with chap enabled on it.
+            # we can do an online copy, which is a background process
+            # on the 3PAR that makes the volume instantly available.
+            # We can't resize a volume, while it's being copied.
+            if dst_volume['size'] == src_vref['size'] and not \
+                    (vol_chap_enabled):
+                LOG.debug("Creating a clone of volume, using online copy.")
```

Comment:
you can make to LOG.info() since for code tracing it will help if the logging is restrictive.